webui/react/src/components/NotebookModal.tsx

```diff
@@ -0,0 +1,309 @@
+    storage.getWithDefault(STORAGE_KEY, { slots: 1 }),
+  );
+
+  const storeConfig = useCallback((values: NotebookConfig) => {
+    delete values.name;
+    storage.set(STORAGE_KEY, values);
+  }, [ storage ]);
+
+  useEffect(() => {
+    storeConfig(state);
+  }, [ state, storeConfig ]);
+
+  return [ state, dispatch ];
+};
+
+interface NotebookModalProps extends ModalProps {
+  onLaunch?: () => void;
+  visible?: boolean;
+}
+
+interface FormProps {
+  fields: NotebookConfig;
+  onChange: DispatchFunction;
+}
+
+interface FullConfigProps {
+  config: RawJson;
+  onChange: (config: RawJson) => void;
+}
+
+interface ResourceInfo {
+  hasCPU: boolean;
+  hasGPU: boolean;
+  showResourceType: boolean;
+}
+
+const NotebookModal: React.FC<NotebookModalProps> = (
+  { visible = false, onLaunch, ...props }: NotebookModalProps,
+) => {
+
+  const [ showFullConfig, setShowFullConfig ] = useState(false);
+  const [ fields, dispatch ] = useNotebookForm();
+  const [ config, setConfig ] = useState<RawJson>({});
+
+  const fetchConfig = useCallback(async () => {
+    try {
+      const config = await previewNotebook(
+        fields.slots,
+        fields.template,
+        fields.name,
+        fields.pool,
+      );
+      setConfig({ config: yaml.dump(config) });
+
+    } catch {}
+  }, [ fields ]);
+
+  useEffect(() => {
+    if (showFullConfig) fetchConfig();
+  }, [ showFullConfig, fetchConfig ]);
+
+  const handleSecondary = useCallback(() => {
+    setShowFullConfig(show => !show);
+  }, [ ]);
+
+  const handleCreateEnvironment = useCallback(
+    () => {
+      if (showFullConfig) {
+        launchNotebook(config);
+      } else {
+        launchNotebook(
+          undefined,
+          fields.slots,
+          fields.template,
+          fields.name,
+          fields.pool,
+        );
+      }
+      if (onLaunch) onLaunch();
+    },
+    [ showFullConfig, onLaunch, fields, config ],
+  );
+
+  return <Modal
+    footer={<>
+      <Button onClick={handleSecondary}>{showFullConfig ? 'Edit Form' : 'Edit Full Config'}</Button>
+      <Button
+        type="primary"
+        onClick={handleCreateEnvironment}>Launch</Button>
+    </>}
+    title="Launch JupyterLab"
+    visible={visible}
+    width={540}
+    {...props}>
+    {showFullConfig ?
+      <NotebookFullConfig config={config} onChange={(newConfig) => setConfig(newConfig)} /> :
+      <NotebookForm fields={fields} onChange={dispatch} />
+    }
+  </Modal>;
+};
+
+const NotebookFullConfig:React.FC<FullConfigProps> = (
+  { config, onChange }:FullConfigProps,
+) => {
+  const [ field, setField ] = useState([ { name: 'config', value: '' } ]);
+
+  useEffect(() => {
+    setField([ { name: 'config', value: yaml.dump(config) } ]);
+  }, [ config ]);
+
+  const extractConfig = useCallback((fieldData) => {
+    onChange(JSON.parse(fieldData[0].value));
+  }, [ onChange ]);
+
+  return <Form
+    fields={field}
+    onFieldsChange={(_, allFields) => {
+      extractConfig(allFields);
+    }}>
```

Comment:
**nit:** just point this `extractConfig` and rename `extractConfig` to something like `handleConfigChange` and that function can look something like:
```
const handleConfigChange = useCallback((_, allFields) => {
  if (!Array.isArray(allFields) || allFields.length === 0) return;
  const configString = allFields[0].value;
  try {
    const config = yaml.dump(configString);
    onChange(config);
  } catch (e) {}
}, []);
```
```